mono/metadata/threads.c

```diff
@@ -5014,12 +5075,13 @@ mono_threads_add_joinable_thread (gpointer tid)
 	joinable_threads_lock ();
 	if (!joinable_threads)
 		joinable_threads = g_hash_table_new (NULL, NULL);
-	g_hash_table_insert (joinable_threads, tid, tid);
-	UnlockedIncrement (&joinable_thread_count);
+	if (!g_hash_table_lookup (joinable_threads, tid)) {
+		threads_add_joinable_thread_nolock (tid);
```

Comment:
Why is the lookup needed ?

Reply:
On Windows we need to store a handle to the thread in the joinable_threads table (as value) to make sure the tid still represents the same thread when waited upon. As long as there is an outstanding handle to a thread the tid won't be reused by the OS. This check prevents that we replace a tid handle, since the handle will then be leaked on Windows platforms. An alternative could be to implement a custom value_destroy_func on Windows to capture this case. Maybe that could be a better solution since it makes the remove scenario cleaner as well. I will update PR with that solution. Thanks!

Reply:
Looked a little more into this, current implementations call of g_hash_table_insert is problematic in cases where the tid is already in the table, since the joinable_thread_count will then not be correct, and there is no way to check if g_hash_table_insert added a new item or just changing previous item in the table. Doing the value_destroy_func on Windows won't solve this, and the problem with the increment of joinable_thread_count exists on other platforms. In previous usage of joinable threads this use case probably didn't happen (same tid added multiple times to table), but it can happen with this fix, so doing the check prevents duplicates, makes sure the joinable threads in the list corresponds to the joinable_thread_count. It also makes the implementation of mono_threads_add_joinable_thread more robust for future use. OK to keep this change as is?

Reply:
Sounds good to me

Reply:
Please use `g_hash_table_lookup_extended` so a `NULL` value isn't considered false.

Reply:
will do